// relay-compiler/lib/index.js
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 * @format
 */
'use strict';

var ASTCache = require('./core/ASTCache');

var ASTConvert = require('./core/ASTConvert');

var CodeMarker = require('./util/CodeMarker');

var CodegenDirectory = require('./codegen/CodegenDirectory');

var CodegenRunner = require('./codegen/CodegenRunner');

var CodegenWatcher = require('./codegen/CodegenWatcher');

var DotGraphQLParser = require('./core/DotGraphQLParser');

var FindGraphQLTags = require('./language/javascript/FindGraphQLTags');

var GraphQLCompilerContext = require('./core/GraphQLCompilerContext');

var GraphQLCompilerProfiler = require('./core/GraphQLCompilerProfiler');

var GraphQLConsoleReporter = require('./reporters/GraphQLConsoleReporter');

var GraphQLIRPrinter = require('./core/GraphQLIRPrinter');

var GraphQLIRTransformer = require('./core/GraphQLIRTransformer');

var GraphQLIRVisitor = require('./core/GraphQLIRVisitor');

var GraphQLMultiReporter = require('./reporters/GraphQLMultiReporter');

var GraphQLSchemaUtils = require('./core/GraphQLSchemaUtils');

var GraphQLWatchmanClient = require('./core/GraphQLWatchmanClient');

var RelayCodeGenerator = require('./codegen/RelayCodeGenerator');

var RelayFileWriter = require('./codegen/RelayFileWriter');

var RelayFlowGenerator = require('./language/javascript/RelayFlowGenerator');

var RelayIRTransforms = require('./core/RelayIRTransforms');

var RelayIRValidations = require('./core/RelayIRValidations');

var RelayParser = require('./core/RelayParser');

var RelaySourceModuleParser = require('./core/RelaySourceModuleParser');

var RelayValidator = require('./core/RelayValidator');

var Rollout = require('./util/Rollout');

var compileRelayArtifacts = require('./codegen/compileRelayArtifacts');

var filterContextForNode = require('./core/filterContextForNode');

var formatGeneratedModule = require('./language/javascript/formatGeneratedModule');

var getIdentifierForArgumentValue = require('./core/getIdentifierForArgumentValue');

var getLiteralArgumentValues = require('./core/getLiteralArgumentValues');

var getNormalizationOperationName = require('./core/getNormalizationOperationName');

var isEquivalentType = require('./core/isEquivalentType');

var nullthrows = require('./util/nullthrowsOSS');

var writeRelayGeneratedFile = require('./codegen/writeRelayGeneratedFile');

var _require = require('./bin/RelayCompilerMain'),
    main = _require.main;

var _require2 = require('./codegen/SourceControl'),
    SourceControlMercurial = _require2.SourceControlMercurial;

var _require3 = require('./core/GraphQLDerivedFromMetadata'),
    getReaderSourceDefinitionName = _require3.getReaderSourceDefinitionName,
    getSourceDefinitionName = _require3.getSourceDefinitionName;

var RelayJSModuleParser = RelaySourceModuleParser(FindGraphQLTags.find);
module.exports = {
  relayCompiler: main,
  ASTConvert: ASTConvert,
  CodegenDirectory: CodegenDirectory,
  CodegenRunner: CodegenRunner,
  CodegenWatcher: CodegenWatcher,
  CodeMarker: CodeMarker,
  CompilerContext: GraphQLCompilerContext,
  ConsoleReporter: GraphQLConsoleReporter,
  DotGraphQLParser: DotGraphQLParser,
  ASTCache: ASTCache,
  IRTransformer: GraphQLIRTransformer,
  IRVisitor: GraphQLIRVisitor,
  Printer: GraphQLIRPrinter,
  Profiler: GraphQLCompilerProfiler,
  Rollout: Rollout,
  SchemaUtils: GraphQLSchemaUtils,
  SourceControlMercurial: SourceControlMercurial,
  WatchmanClient: GraphQLWatchmanClient,
  filterContextForNode: filterContextForNode,
  getIdentifierForArgumentValue: getIdentifierForArgumentValue,
  getNormalizationOperationName: getNormalizationOperationName,
  getLiteralArgumentValues: getLiteralArgumentValues,
  isEquivalentType: isEquivalentType,
  nullthrows: nullthrows,
  Parser: RelayParser,
  Validator: RelayValidator,
  CodeGenerator: RelayCodeGenerator,
  FlowGenerator: RelayFlowGenerator,
  GraphQLCompilerContext: GraphQLCompilerContext,
  FileWriter: RelayFileWriter,
  IRTransforms: RelayIRTransforms,
  IRValidations: RelayIRValidations,
  JSModuleParser: RelayJSModuleParser,
  MultiReporter: GraphQLMultiReporter,
  Runner: CodegenRunner,
  compileRelayArtifacts: compileRelayArtifacts,
  formatGeneratedModule: formatGeneratedModule,
  convertASTDocuments: ASTConvert.convertASTDocuments,
  transformASTSchema: ASTConvert.transformASTSchema,
  getReaderSourceDefinitionName: getReaderSourceDefinitionName,
  getSourceDefinitionName: getSourceDefinitionName,
  writeRelayGeneratedFile: writeRelayGeneratedFile
};
